legate-boost tutorial

NihaalChowdary opened the pull request:
legate-boost tutorial doc with code files

NihaalChowdary committed:
Add files via upload

NihaalChowdary committed:
Update Legate_Boost.rst

NihaalChowdary committed:
Add files via upload

NihaalChowdary committed:
Update Legate_Boost2.rst

NihaalChowdary committed:
Update Legate_Boost2.rst

NihaalChowdary committed:
Update Legate_Boost2.rst

NihaalChowdary committed:
Delete docs/source/Legate_Boost.rst

NihaalChowdary committed:
Rename Legate_Boost2.rst to Legate_Boost.rst

NihaalChowdary committed:
Update Legate_Boost.rst

NihaalChowdary committed:
Add files via upload

NihaalChowdary committed:
Add files via upload

NihaalChowdary committed:
Update creditscore.py

NihaalChowdary committed:
Update housing.py

NihaalChowdary committed:
Update Legate_Boost.rst

NihaalChowdary committed:
Update Legate_Boost.rst

NihaalChowdary committed:
Update Legate_Boost.rst

NihaalChowdary committed:
Update housing.py

NihaalChowdary committed:
Update Legate_Boost.rst

NihaalChowdary committed:
Update creditscore.py

NihaalChowdary committed:
Update Legate_Boost.rst

NihaalChowdary committed:
Update creditscore.py

NihaalChowdary committed:
Update housing.py

NihaalChowdary committed:
Update Legate_Boost.rst

NihaalChowdary committed:
Add files via upload

NihaalChowdary committed:
Add files via upload

NihaalChowdary committed:
Update inference.py

NihaalChowdary committed:
Update LBClassifier.py

NihaalChowdary committed:
Update Legate_Boost.rst

NihaalChowdary committed:
Update creditscore.py

NihaalChowdary committed:
Update Legate_Boost.rst

NihaalChowdary committed:
Update Legate_Boost.rst

NihaalChowdary committed:
Update Legate_Boost.rst

NihaalChowdary committed:
Update Legate_Boost.rst

NihaalChowdary committed:
Update legate_boost.rst

NihaalChowdary committed:
Rename Legate_Boost.rst to legate-boost.rst

NihaalChowdary committed:
Update legate-boost.rst

RAMitchell commented:
/ok to test

RAMitchell reviewed (commented):
Thank you for adding documentation!

The main thing we need here is tests added to https://github.com/rapidsai/legate-boost/blob/main/legateboost/test/test_examples.py that can run the examples in CI.

Without testing, code examples go out of date which can be quite frustrating for users.

NihaalChowdary committed:
Update legate-boost.rst

NihaalChowdary committed:
removed installation steps and liked to repo

NihaalChowdary committed:
Updated installation links

NihaalChowdary commented:
Hi @RAMitchell,
Would you like me to add the tutorial’s code files to the [examples directory](https://github.com/rapidsai/legate-boost/tree/main/examples) and create a new subdirectory for them, so they can be included in the CI tests?

NihaalChowdary committed:
Rename docs/source/LBClassifier.py to docs/source/examples/LBClassifier.py

NihaalChowdary committed:
Rename docs/source/LBRegressor.py to docs/source/examples/LBRegressor.py

NihaalChowdary committed:
Rename docs/source/creditscore.py to docs/source/examples/creditscore.py

NihaalChowdary committed:
Rename docs/source/housing.py to docs/source/examples/housing.py

NihaalChowdary committed:
Rename docs/source/inference.py to docs/source/examples/inference.py

NihaalChowdary committed:
Update legate-boost.rst

NihaalChowdary committed:
Update creditscore.py

NihaalChowdary committed:
Update housing.py

NihaalChowdary committed:
Update housing.py

NihaalChowdary committed:
Update inference.py

NihaalChowdary committed:
Apply pre-commit formatting and linting fixes

NihaalChowdary committed:
Update legate-boost.rst

RAMitchell commented:
Yes I think it makes sense to put them in the example directory. You will see some examples use an environment variable to tell if they are running in CI to reduce the amount of data and the test run-time. Can you also please check whether any of your examples directly overlap the existing examples, then we can perhaps remove some old examples.

NihaalChowdary committed:
updated data for CI

NihaalChowdary committed:
Updated data for CI

NihaalChowdary committed:
Updated data and model for CI

NihaalChowdary committed:
Move tutorial examples to examples directory for CI

NihaalChowdary committed:
Updated the code files path

NihaalChowdary commented:
Hi @RAMitchell,
I have moved the code files to the examples/tutorial_examples folder. I also modified the code to select a smaller dataset for CI runs. I have checked, and I don’t see any examples that overlap in terms of data usage, logic, or problem execution.

NihaalChowdary committed:
Update legate-boost.rst

RAMitchell commented:
/ok to test d58f7a6f17a8a0f87dea452527d203c10e50f795

RAMitchell committed:
Fix pre-commit

Signed-off-by: Rory Mitchell <r.a.mitchell.nz@gmail.com>

RAMitchell commented on docs/source/legate-boost.rst:
```diff
@@ -0,0 +1,694 @@
+
+.. code-block:: sh
+
+   legate --cpus 1 --gpus 0 ./inference.py
```
Personally I would consider multiple cpus (8-16) against 1 gpu as a more fair test.

NihaalChowdary replied:
Hi @RAMitchell,

Since the model size is relatively small, the computation is negligible when increasing the number of CPUs. The timings with 1 CPU and 16 CPUs are very similar, so showing results with 16 CPUs might be misleading:

1 CPU
Mean: 265.66 ms
Median: 262.97 ms
Min: 249.78 ms
Max: 284.44 ms
Variance: 117,319.15 ms
Std. Dev.: 10.83 ms

16 CPUs
Mean: 247.65 ms
Median: 248.21 ms
Min: 238.37 ms
Max: 255.27 ms
Variance: 27,581.68 ms
Std. Dev.: 5.25 ms

This shows that parallelism has little effect in this case, and using 1 CPU is sufficient for the demonstration.

RAMitchell reviewed (commented):
I will fix the pre-commit errors, many of which are not related to this PR.

legate-boost.rst needs to be included in index.rst so sphinx knows to build it.

Have you been able to check the docs build locally? `make html` from the docs folder and then you can use a web server to preview the html pages.

Legate-boost currently has no dependency on legate-dataframe so I will have to add that for the tests to pass.

Generally speaking the examples look great. I don't think people will take seriously the comparison of 1 cpu core against a gpu for inference. I suspect most of the time here is latency inherent in legate and cupynumeric anyway - 1 gpu should be able to predict orders of magnitude faster than 1 cpu core, there just isn't enough work in this task to saturate the GPU. You can still keep all the timing code, I would just bump the cpus to 8 and then mention that this task is not compute bound. You will probably see a dramatic speed-up of GPU vs CPU during training with a neural network model if that is what you want to show.

NihaalChowdary committed:
Update legate-boost.rst

NihaalChowdary commented on docs/source/legate-boost.rst:
```diff
@@ -0,0 +1,694 @@
+
+.. code-block:: sh
+
+   legate --cpus 1 --gpus 0 ./creditscore.py
```
bumped to 8 cpus

NihaalChowdary committed:
Update legate-boost.rst

NihaalChowdary committed:
Add files via upload

NihaalChowdary committed:
Update inference.py

NihaalChowdary committed:
Update legate-boost.rst

NihaalChowdary commented on docs/source/legate-boost.rst:
```diff
@@ -0,0 +1,694 @@
+
+      # [import libraries]
+      import os
+      import cudf
```
Understood. So for this project, do we want to rely only on pandas and not use cudf at all?

RAMitchell commented on docs/source/legate-boost.rst:
```diff
@@ -0,0 +1,694 @@
+   rt = lg.get_legate_runtime()
+
+   timings = []
+   model = load("legate_boost_housing.joblib")
```
Uploading binaries to git is generally not a good idea. You would instead need the code that generated the file somewhere. Keep in mind that CI will try to run every block of code it finds and fail if it can't run it.

NihaalChowdary replied:
# [Save model]
dump(model, "legate_boost_model.joblib")

# [Save test data]
x_test_cpu = x_test.get() if hasattr(x_test, "get") else np.array(x_test)
y_test_cpu = y_test.get() if hasattr(y_test, "get") else np.array(y_test)

pd.DataFrame(x_test_cpu).to_csv("x_test.csv", index=False)
pd.DataFrame(y_test_cpu, columns=["Target"]).to_csv("y_test.csv", index=False)

In the creditscore.py file, this section of the code generates the model and test dataset, then saves them for later use. The resulting model and dataset files are subsequently used by inference.py to perform the inference calculations.  will it not pass in the CI ? In that case can i do this:

if not (os.path.exists("legate_boost_model.joblib") and 
        os.path.exists("x_test.csv") and 
        os.path.exists("y_test.csv")):
    print("Required files not found. Regenerating using creditscore.py logic...")

    from tutorial_examples.creditscore import model, x_test, y_test  # reuse training code

    # Save model
    dump(model, "legate_boost_model.joblib")

    # Save test data
    x_test_cpu = x_test.get() if hasattr(x_test, "get") else np.array(x_test)
    y_test_cpu = y_test.get() if hasattr(y_test, "get") else np.array(y_test)

    pd.DataFrame(x_test_cpu).to_csv("x_test.csv", index=False)
    pd.DataFrame(y_test_cpu, columns=["Target"]).to_csv("y_test.csv", index=False)

# load files 
model = load("legate_boost_model.joblib")
X = pd.read_csv("x_test.csv")

NihaalChowdary committed:
Update creditscore.py

RAMitchell committed:
Merge branch 'main' of github.com:rapidsai/legateboost into tutorial

RAMitchell committed:
Fix docs build, remove binary files

Signed-off-by: Rory Mitchell <r.a.mitchell.nz@gmail.com>

RAMitchell commented:
I think I've solved the other unrelated CI issues, and have removed the binary files.

I added the new rst file to the index and built the docs locally and they look good!

The tests that run the examples fail currently. You can run them locally like
`legate --module pytest legateboost/test/test_examples.py`

Please have a go at fixing these up.

RAMitchell commented on docs/source/legate-boost.rst:
```diff
@@ -0,0 +1,441 @@
+and `legate boost`_ for more
+information and detailed instructions on installation and setup.
+
+.. _Distributed Computing with cupynumeric: https://github.com/NVIDIA/accelerated-computing-hub/blob/main/Accelerated_Python_User_Guide/notebooks/Chapter_11_Distributed_Computing_cuPyNumeric.ipynb
```
This link will 404 if somebody moves the file in this repo. So we either need to check the links don't 404 in CI somehow or link to something that is more likely to stay put.

RAMitchell commented:
/ok to test b90ef3438f1920173241228cfdf49a8d9f09871f

RAMitchell committed:
Pre-commit

Signed-off-by: Rory Mitchell <r.a.mitchell.nz@gmail.com>

NihaalChowdary committed:
Update creditscore.py

NihaalChowdary committed:
Update legate-boost.rst

NihaalChowdary committed:
merged creditscore.py and inference.py

NihaalChowdary committed:
Delete examples/tutorial_examples/inference.py

NihaalChowdary committed:
Updated the link to cupynumeric tutorial

NihaalChowdary committed:
Update docs and tutorial examples

RAMitchell commented:
/ok to test 07f19b2b1f02fe150bb3dda88c9c36fa7a34264a

dongb commented:
Hi Rory, any update on this doc publishing?
thanks
-bo

NihaalChowdary committed:
Merge branch 'rapidsai:main' into main
